Feature: user configurable columns

ElPollero opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 7.x)
* a.b for any bug fixes (e.g. 5.2, 6.0)
* c.x for any bug fixes, features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 7.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     |  <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->  I came across this idea and decided to implement it. This PR adds user-configurable table columns in views as discussed here. https://forum.mautic.org/t/user-configurable-table-columns-in-views/31850

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

This PR introduces a framework that enables users to customize the visibility of columns in admin tables. Currently, it is implemented for the Companies, Contacts, and Emails tables, where a new "Columns" button is displayed in the toolbar. Clicking this button opens a modal with checkboxes for each available column, allowing users to tailor the table view to their preferences.

User preferences are saved via an AJAX request to /s/user/set-preference and persist across sessions using Mautic’s native preference system. 

The feature is built as a generic and extensible framework. The 16columns.js file contains a window.columnConfigs object where each supported table's configuration is defined. To extend this feature to additional tables, developers simply need to add the relevant configuration for the new table into that object.

Currently, the "Columns" button is restricted to only the tables for which configuration is available, in order to avoid interfering with unconfigured table pages. However, it can easily be made universal by removing this restriction, making the integration of future tables straightforward and low-effort.

Includes functional tests and E2E tests.
Co-authored-by: Pedro Loureiro <pedro.d.loureiro@tecnico.ulisboa.pt>

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1.Go to /s/companies, /s/contacts, or /s/emails.

2. Reload the page and click the new “Columns” button in the top-right toolbar.

3.A modal will open with checkboxes for each column.

4.Toggle columns on or off, then click "Save changes".

5.The table updates based on selected columns.

6.Preferences are stored individually per user.

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

ElPollero committed:
feat: implementing user configurable columns for email page

- Added "columns" button to toggle visibility of email list columns
- Implemented saving of user column
  preferences via AJAX to user settings
- Applied `userColumnPrefs` consistently after table
  operations like sorting and pagination
Co-authored-by: Pedro Loureiro <pedro.d.loureiro@tecnico.ulisboa.pt>

ElPollero committed:
feat: Implement configurable column visibility framework for tables

Introduced a reusable and extensible JS-based system for
user-configurable column visibility across key data tables
(Contacts, Companies, Emails).

- Added `window.columnConfigs` to centralize per-route table metadata
 (selectors, column class prefixes, labels, preference keys).

- Created `applyColumnPrefs()` and `applyColumnPrefsForRoute()`
 utilities to apply visibility toggles based on stored preferences.

- Replaced hardcoded column checkboxes with dynamic modal rendering
via `renderColumnModal()` using route-specific config.

- Updated Mautic.reorderTableData, filterTableData, limitTableData,
 and filterList to invoke applyColumnPrefsForRoute() after content
reloads.

- Refactored controllers (`CompanyController`, `LeadController`)
 to pass stored preferences to Twig views.

- Updated table `td`/`th` classes to follow the required
 `col-<entity>-<field>` convention for dynamic targeting.

- Ensured user preferences are saved via `/s/user/set-preference`
 POST, using `setupColumnPreferences()` per context.

This sets the foundation for future extensibility of list views, and
improves UX by allowing users to tailor visible data columns per
entity type.

Co-authored-by: Pedro Loureiro <pedro.d.loureiro@tecnico.ulisboa.pt>

ElPollero committed:
feat: Add tests for user column preferences

- Implemented e2e acceptance test (ColumnPreferencesCest)
  verifying company column visibility toggling in UI
    * Tests login, company creation, and column preference selection
    * Includes screenshot verification at key steps
    * Verifies DOM elements visibility after preferences are applied

- Added functional test (UserPreferenceTest) for user preference API
    * Tests setting user preferences via POST request
    * Verifies successful response and proper database storage
    * Includes helper method to create test users with admin role

- Both test suites cover complementary aspects of preference system:
    * UI workflow through acceptance tests
    * API and backend functionality through functional tests
Co-authored-by: Pedro Loureiro <pedro.d.loureiro@tecnico.ulisboa.pt>

PedroLoureiro1 committed:
fix: implement configurable column preferences for core modules

- "Columns" button restricted to implemented modules only
  (Contacts, Companies, Emails)
  * It can be easily removed so the button appears on every table
- Implemented seamless preference saving via optimized AJAX calls
- Refactored modal handling for smoother save/apply workflow:
  * Fixed preference persistence between sessions
  * Eliminated need for manual refresh after changes

Co-authored-by: Andre Bento <andre.s.bento@tecnico.ulisboa.pt>

PedroLoureiro1 committed:
refactor: remove inline CSS and external CDN includes from Twig

- Moved custom column picker styles from inline <style>
  block to app.css
- Removed jQuery and Bootstrap CDN includes from the Twig template
- Relied on existing globally available jQuery and Bootstrap
  instances provided by the application

Co-authored-by: Andre Bento <andre.s.bento@tecnico.ulisboa.pt>

RCheesley committed:
Merge branch '7.x' into feature/user-configurable-columns

ElPollero commented:
Hi @RCheesley,

Thanks for the heads-up!

Just a quick update on the test failures:

Our E2E test isn’t passing on CI, although it worked fine locally. The test itself might not be related to the recent DB migrations, but I’m currently unable to debug it because I’m running into issues with the new migrations after upgrading this branch from 6.x to 7.x.

PHPStan is also flagging a few minor type mismatch issues. We believe we've addressed them, but we'll double-check to be sure.

If you or anyone else is able to run the E2E tests locally or help debug the issue, that would be greatly appreciated!

Thanks again

PedroLoureiro1 committed:
fix: Add @var note to help PHPStan recognize getPreference method

Added docblock @var annotations to clarify the concrete type of
AbstractCommonModel, allowing PHPStan to resolve the getPreference()
method

Co-authored-by: Andre Bento <andre.s.bento@tecnico.ulisboa.pt>

shinde-rahul commented on app/bundles/CoreBundle/Resources/views/Helper/list_toolbar.html.twig:
```diff
@@ -40,8 +48,191 @@
 
 </div>
+
+<script>
```
Move this script to a separate JS file and load that JavaScript on the pages.

shinde-rahul commented on app/bundles/CoreBundle/Resources/views/Helper/list_toolbar.html.twig:
```diff
@@ -40,8 +48,191 @@
+}
+</script>
+
+<script>
```
Move this script to its file.

PedroLoureiro1 replied:
Hi @shinde-rahul,

I hope you're doing well.

I just wanted to clarify something regarding the snippet you shared. It includes two scripts, and I wasn’t entirely sure which one you were referring to. Were you expecting us to move both scripts into their respective files?

Thanks in advance for the clarification.

RCheesley replied:
@PedroLoureiro1 all scripts should be loaded from separate files, not combined with Twig or PHP etc.

shinde-rahul commented on app/bundles/EmailBundle/Controller/EmailController.php:
```diff
@@ -230,19 +231,24 @@ public function indexAction(Request $request, EmailModel $model, EmailConfig $em
         }
         $session->set('mautic.email.page', $page);
 
+        /** @var UserModel $userModel */
+        $userModel   = $this->getModel('user');
```
```suggestion

```
This has already been added as an explicit dependency to the method.

shinde-rahul commented on app/bundles/LeadBundle/Controller/CompanyController.php:
```diff
@@ -96,17 +96,22 @@ public function indexAction(Request $request, PageHelperFactoryInterface $pageHe
         $companyIds = array_keys($companies);
         $leadCounts = (!empty($companyIds)) ? $model->getRepository()->getLeadCount($companyIds) : [];
 
+        /** @var \Mautic\UserBundle\Model\UserModel $userModel */
+        $userModel           = $this->getModel('user');
```
Add `UserModel  user$userModel` as an explicit dependency to the `indexAction` method.

shinde-rahul commented on app/bundles/LeadBundle/Controller/LeadController.php:
```diff
@@ -211,6 +212,10 @@ public function indexAction(
         \assert($leadDNCModel instanceof DoNotContactModel);
         $dncRepository = $leadDNCModel->getDncRepo();
 
+        /** @var UserModel $userModel */
+        $userModel            = $this->getModel('user');
```
Add `UserModel  user$userModel` as an explicit dependency to the `indexAction` method.

shinde-rahul reviewed (changes requested):
There are a few issues that need attention. Please check the SonarQube https://sonarcloud.io/dashboard?id=mautic_mautic&pullRequest=15097

escopecz commented on app/bundles/CoreBundle/Assets/css/app.css:
```diff
@@ -11351,3 +11351,32 @@ code:hover > .copy-icon {
 .contained-list-item__action > * {
   pointer-events: all;
 }
+/* === Column Picker Modal Styles === */
+#column-picker-form {
+  max-width: 400px;
+  margin: 0 auto;
+}
+
+#column-picker-form .form-check {
+  display: flex;
+  justify-content: space-between;
+  align-items: center;
+  margin-bottom: 12px;
+  padding: 6px 0;
+  font-size: 1.3rem;
+}
+
+#column-picker-form .form-check-label {
+  margin-bottom: 0;
+  flex-grow: 1;
+  cursor: pointer;
+  font-weight: 600;
+}
+
+#column-picker-form .form-check-input {
+  margin-left: 30px;
+  flex-shrink: 0;
+  width: 24px;
+  height: 24px;
+  cursor: pointer;
+}
```
The app.css file is generated from the less files. So this must be moved there otherwise it will be removed next time someone will rebuild this file. See https://contribute.mautic.org/contributing-to-mautic/developer/local-environment-setup#development-environment.

escopecz commented on app/bundles/CoreBundle/Assets/js/16.columns.js:
```diff
@@ -0,0 +1,67 @@
+window.columnConfigs = {
```
I'd suggest to have this config in a PHP file so even plugins could hook into this feature. Each bundle can now add JS during the `BUILD_MAUTIC_JS` event so that could be used here as well.

escopecz commented on app/bundles/UserBundle/Controller/ProfileController.php:
```diff
@@ -237,4 +239,18 @@ public function indexAction(Request $request, LanguageHelper $languageHelper, Us
+
+        $userModel->setPreference($key, $value, $user);
+
+        return new JsonResponse(['success' => true]);
```
I know that some old controllers respond with the success message but it's a duplicated information. That's what the 200 response code is for. Let's remove it and respond with empty JSON array if nothing else is needed.

escopecz commented on app/bundles/UserBundle/Tests/Functional/UserPreferenceTest.php:
```diff
@@ -0,0 +1,79 @@
+{
+    public function testUserCanSetPreference(): void
+    {
+        $user = $this->createTestUser();
+
+        $this->loginUser($user);
+        $this->client->setServerParameter('PHP_AUTH_USER', $user->getUsername());
+        $this->client->setServerParameter('PHP_AUTH_PW', 'Maut1cR0cks!');
```
Is this necessary? The admin user is already created and logged in for every functional test.

escopecz commented on app/bundles/UserBundle/Tests/Functional/UserPreferenceTest.php:
```diff
@@ -0,0 +1,79 @@
+
+        $response = $this->client->getResponse();
+
+        $this->assertEquals(Response::HTTP_OK, $response->getStatusCode(), $response->getContent());
```
```suggestion
        $this->assertResponseIsSuccessful();
```
this does the same thing and adds some more debug info if fails.

escopecz commented on app/bundles/CoreBundle/Resources/views/Helper/list_toolbar.html.twig:
```diff
@@ -40,8 +48,191 @@
+  columnPrefix,
+  csrfToken
+}) {
+  const $ = window.jQuery;
```
Please use `mQuery` which is already a Mautic's alias for jQuery.

escopecz commented on app/bundles/CoreBundle/Resources/views/Helper/list_toolbar.html.twig:
```diff
@@ -40,8 +48,191 @@
+    document.querySelectorAll('[aria-hidden="true"]').forEach(el => el.removeAttribute('aria-hidden'));
+  });
+
+  $('#column-config-modal').on('shown.bs.modal', function () {
```
If you are using the same jQuery element twice, please store it to a variable and reuse it. Creating the element every time from a selector is not cheap.

escopecz commented on tests/acceptance/ColumnPreferencesCest.php:
```diff
@@ -0,0 +1,67 @@
+        $I->click('#col-score');
+        $I->click('#col-contacts');
+        $I->click('#col-id');
+        $I->wait(0.5);
```
This seem sketchy. It may cause a flaky test. Can we wait for some element to appear instead?

PedroLoureiro1 committed:
refactor: modularize column picker styles, improve accessibility

- Removed inline CSS from app.css, migrated to modular
LESS import (components/columnpicker.less)

- Replaced global window.columnConfigs with
window.MauticColumnPrefs IIFE module:

 ->Centralized column config management

 ->Added error handling for invalid parameters

 ->Applied preferences on DOM load and Turbolinks navigation

- Improved modal accessibility:

  ->Correct handling of aria-hidden

  ->Simplified modal event bindings
(shown.bs.modal, hidden.bs.modal)

- Cleaned up Twig templates:

  ->Removed inline applyColumnPrefs

  ->Robust default initialization for column preferences

- Refactored controllers:

  ->Injected dependencies (e.g., UserModel, CompanyModel)
instead of $this->getModel()

Co-authored-by: Andre Bento andre.s.bento@tecnico.ulisboa.pt

PedroLoureiro1 committed:
refactor: streamline column preferences and test infrastructure

- Removed redundant parent() call in list_toolbar.html.twig:
  → Prevents duplicate script execution
  → Cleaner template inheritance

- Refactored CompanyController dependencies:
  → Explicitly injected UserModel via constructor
  → Improved method signature formatting
  → Replaced $this->getModel() with DI

- Optimized UserPreferenceTest:
  → Replaced undefined UserProvider with UserModel
  → Removed unused createTestUser() method
  → Simplified using default admin user (ID 1)

Co-authored-by: Andre Bento andre.s.bento@tecnico.ulisboa.pt
